File: aplicacao/__brick__/lib/data/storage/storage.dart
import 'package:dartz/dartz.dart';
import 'package:my_application_name/data/data.dart';

export 'file_storage.dart';
export 'preferences_storage.dart';

abstract class Storage<I, O> {
  Future<Either<RequestError, O>> store({
    required final String key,
    required final I data,
  });

  Future<Either<RequestError, O>> lookup({
    required final String key,
  });

  Future<Either<RequestError, bool>> exists({
    required final String key,
  });

  Future<Either<RequestError, void>> evict({
    required final String key,
  });
}

File: aplicacao_web/__brick__/src/lib/data/storage/storage.ts
import type { Either, IOError } from '@web-pacotes/foundation-types';

/**
 * A general purpose interface for accessing a data storage. The generic types [I] and [O] mark
 * the input and output types of the storage.
 */
export interface Storage<I, O> {
	store(key: string, data: I): Promise<Either<IOError, O>>;

	lookup(key: string): Promise<Either<IOError, O>>;

	exists(key: string): Promise<Either<IOError, boolean>>;

	evict(key: string): Promise<Either<IOError, void>>;
}
